js/compiler.ts

```diff
@@ -495,11 +509,71 @@ export class Compiler
       };
     });
   }
+}
 
-  private static _instance: Compiler | undefined;
+const compiler = new Compiler();
 
-  /** Returns the instance of `Compiler` or creates a new instance. */
-  static instance(): Compiler {
-    return Compiler._instance || (Compiler._instance = new Compiler());
+let startResMsg: msg.StartRes;
+
+/** Send to the privilaged side that we have setup and are ready. */
+function sendStart(): void {
+  const builder = flatbuffers.createBuilder();
+  msg.Start.startStart(builder);
+  const startOffset = msg.Start.endStart(builder);
+  const baseRes = sendSync(builder, msg.Any.Start, startOffset);
+  assert(baseRes != null);
+  assert(msg.Any.StartRes === baseRes!.innerType());
+  startResMsg = new msg.StartRes();
+  assert(baseRes!.inner(startResMsg) != null);
+}
+
+// set global objects for compiler web worker
+window.clearTimeout = clearTimer;
+window.console = console;
+window.postMessage = postMessage;
+window.setTimeout = setTimeout;
+window.workerMain = workerMain;
+window.close = workerClose;
+window.TextDecoder = TextDecoder;
+window.TextEncoder = TextEncoder;
+
+// provide the "main" function that will be called by the privilaged side when
+// lazy instantiating the compiler web worker
+window.compilerMain = function compilerMain() {
+  // workerMain should have already been called since a compiler is a worker.
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+  compiler.recompile = startResMsg.recompileFlag();
+  log(`recompile ${compiler.recompile}`);
+  window.onmessage = ({ data }: { data: Uint8Array }) => {
+    const json = decoder.decode(data);
+    const lookup = JSON.parse(json) as CompilerLookup;
+
+    const moduleMetaData = compiler.compile(lookup.specifier, lookup.referrer);
+
+    const responseJson = JSON.stringify(moduleMetaData);
+    const response = encoder.encode(responseJson);
+    postMessage(response);
+  };
+};
+
+/* tslint:disable-next-line:no-default-export */
+export default function denoMain() {
```

Comment:
Hm.. It would be good if both sides shared this code... But maybe that's difficult to do at this stage.
Could you add a comment about this? 
```
// TODO In the future it would be good if main.ts and this shared code for denoMain instead of having separate implementations. 
```

Reply:
I DRY'ed what I could and put it in `os.start()`.  Feels a lot cleaner.